lib/Conversion/TritonGPUToLLVM/ConvertLayoutOpToLLVM/SharedToDotOperandMFMA.cpp

```diff
@@ -41,16 +41,14 @@ Type getShemPtrTy(Type elemTy) {
   return ptr_ty(elemTy, 3);
 }
 
-// Get a waveId for M axis.
-Value getWaveM(ConversionPatternRewriter &rewriter, Location loc, Value wave,
-               const ArrayRef<unsigned int> &wpt, int elemPerInstr, int M) {
-  return urem(urem(wave, i32_val(wpt[0])), i32_val(M / elemPerInstr));
-}
-// Get a waveId for N axis.
-Value getWaveN(ConversionPatternRewriter &rewriter, Location loc, Value wave,
-               const ArrayRef<unsigned int> &wpt, int elemPerInstr, int N) {
-  Value waveMN = udiv(wave, i32_val(wpt[0]));
-  return urem(urem(waveMN, i32_val(wpt[1])), i32_val(N / elemPerInstr));
+// Get waveId inside block of waves.
+Value getWaveIdInBlock(ConversionPatternRewriter &rewriter, Location loc,
+                       Value waveId, const ArrayRef<unsigned int> &wpt,
+                       int elemPerInstrNonK, int tensorSizeNonK, int nonKIdx) {
+  if (nonKIdx == 1)
+    waveId = udiv(waveId, i32_val(wpt[0]));
+  return urem(urem(waveId, i32_val(wpt[nonKIdx])),
+              i32_val(tensorSizeNonK / elemPerInstrNonK));
```

Comment:
I'm confused about this part.
Say we have warpsPerCTA={2,2}, waveId=1, then we are talking about the top left wave in the workgroup, right?
If so, we have
- for opA, i.e. nonKIdx=0, spatialWarpId = (waveId % wpt[0]) % (M / 32) = 1
- for opB, i.e. nonKIdx=1, spatialWarpId = ((waveId/wpt[0])%wpt[1]) % (N / 8) = 0

But shouldn't wave1 has index 1 for opB and index 0 for opA?
![example](https://github.com/ROCmSoftwarePlatform/triton/assets/3949632/819e6a7d-22c5-46de-ab85-6569ca6f58f6)

Reply:
> But shouldn't wave1 has index 1 for opB and index 0 for opA?

It was originally implemented like this. I am not sure, if one or another orientation have advantages. I did not try other layout.

If you think transposed wave indexing could have advantages or preferable for style reasons, we can swap it and see what happens.

Reply:
so it's assumed that the 2x2 layout is 

wave0       wave2
wave1       wave3

These two assumptions are not about styles, but correctness. Maybe this is the reason why some gemm tests failed in https://github.com/ROCmSoftwarePlatform/triton/pull/402, in which the mfma layout is used directly for global store.

Reply:
You are probably right. I've investigated failures in 402 a little a few weeks ago, and found that current mfma layout is not compatible with global store implementation.